### PR TITLE
#448: Paths with events

### PR DESCRIPTION
**Pull request by EDsCODE**
Addresses #448 
## Changes
- adds options on paths page to allow different type of paths beyond pageviews

## Checklist
- [ ] All querysets/queries filter by Team (if applicable)
- [ ] Backend tests (if applicable)

### CODE REVIEW

File: posthog/api/paths.py [hunk 1]
```diff
@@ -60,16 +78,33 @@ def list(self, request):
...
-
             urls = []
+
+            if event == "$autocapture":
+                rows_query, params = rows.query.sql_with_params()
+                source_element = 'SELECT e."tag_name" as tag_name_source, e."text" as text_source FROM "posthog_element" e JOIN \
+                    ( SELECT group_id, MIN("posthog_element"."order") as minOrder FROM "posthog_element" GROUP BY group_id) e2 ON e.order = e2.minOrder AND e.group_id = e2.group_id where e.group_id = v1.group_id_1'
+                target_element = 'SELECT e."tag_name" as tag_name_target, e."text" as text_target FROM "posthog_element" e JOIN \
+                    ( SELECT group_id, MIN("posthog_element"."order") as minOrder FROM "posthog_element" GROUP BY group_id) e2 ON e.order = e2.minOrder AND e.group_id = e2.group_id where e.group_id = v1.group_id_2'
+                new_query = 'SELECT * FROM ({}) as v1 JOIN LATERAL ({}) as v2 on true JOIN LATERAL ({}) as v3 on true order by -count'.format(rows_query, source_element, target_element)
```
[the remark this answers is not in the document]
Comment (EDsCODE) [reply posthog/api/paths.py:94]: Is there a better way to join the elements without going through element_group?

[the remark this answers is not in the document]
Comment (timgl) [reply posthog/api/paths.py:94]: I think this is fine (though `source_element` and `target_element` seems super similar?)

However at this point, the `rows` above has already been executed due to `[0: 6]`. Worth making sure that doesn't happen. Also might be worth splitting this function out into a few more functions for easy of readability

File: posthog/api/test/test_paths.py [hunk 2]
```diff
@@ -66,4 +66,116 @@ def test_paths(self):
...
+
+        self.assertEqual(response[2]['sourceLabel'], '1_/pricing')
+        self.assertEqual(response[2]['targetLabel'], '2_/')
+        self.assertEqual(response[2]['value'], 1)
+
+        self.assertEqual(response[3]['sourceLabel'], '2_/pricing', response[3])
+        self.assertEqual(response[3]['targetLabel'], '3_/about')
+        self.assertEqual(response[3]['value'], 1)
+
+    def test_autocapture_paths(self):
```
Comment (EDsCODE) [inline posthog/api/test/test_paths.py:142]: opted to keep this one brief because the logic is tested above

**Comment by EDsCODE** [comment]
waiting on #729

**Comment by EDsCODE** [comment]
@timgl Can be re-reviewed. Found something funny happening with the icon alignment. Seems like the dropdown icons get un-aligned but other antd icons are aligned with the reverse 0.125em added in previous PR

Screen Shot 2020-05-11 at 9 59 05 PM

**Review by timgl — COMMENTED** [review]
I think this is pretty good!

RE the icons being weird, might be a case of merging in master again as it is fixed there.

image
With autocapture there are a lot of events that are missing the text. One thing that might be missing is going up to tree. So if you have something like this:
```html
<a href='blabla'>
<span>some text</span>
</a>
```
I think currently this wouldn't show up.

I also do think it'd be useful if you could somehow click it and show the last 10 events, or maybe show the entire DOM tree. It's hard to get the context of what it means. This might be scope creep for this PR though.

**Comment by EDsCODE** [comment]
RE: missing text
The query I added ensures that always the lowest order element is the one that's "focused on" when creating the element paths.

Screen Shot 2020-05-14 at 10 12 02 AM

Did a little digging and found some strange elements. These elements both appear on the /login path but somehow a div that looks like an input field is a lowest order element. I think these are the elements being picked up in the paths. Any idea why they're being saved like this and how to filter around this?

**Comment by timgl** [comment]
The way posthog-js works is it adds a listener to these elements: `['a', 'button', 'form', 'input', 'select', 'textarea', 'label']`

Any click/submission on any of this saves the entire tree. In the above case, the "form" was clicked, though the top level element in this case was the div as this person mis-clicked.

The way to solve this is by looking for the first element that belongs to any of those "useful elements", then grabbing the text all the way up the chain. In this case, there's no text but it should show "<form> click"

**Comment by EDsCODE** [comment]
Got it. Is there anything else related to going up the tree other than finding the minimum order element in the element group? If the SQL I included is querying as intended it should be returning the lowest order element (top most element) in the group so those tags in the picture might just have no associated text. Will check for correctness of query.

As for providing more details, I'll take a look at how we can display more details/more of the tree for ambiguous elements.

**Comment by EDsCODE** [comment]
@timgl Full DOM display should be good to review and show that the elements even without text are the top of the tree. I opted to use a separate component to show the DOM tree because I dont think it's possible to render in a jsx component in the sankey. One thing not sparking joy is that it can't autoscroll to the top to show elements if you click on a tag because of the layout component we're using. Can investigate this further in a separate issue as it might require a bit of refactoring.

**Comment by timgl** [comment]
Looks great! I would suggest putting those HTML tags in a Modal instead if you can't do scroll by top. It really confused me when I kept clicking and nothing kept happening 😄

**Comment by EDsCODE** [comment]
@timgl ready for another look with modals

**Comment by timgl** [comment]
Looks great!

**Pull request merged**